In [5]:
# Imported Dependencies

import os
import numpy as np
import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, classification_report

In [6]:
# Credit Risk data imported
path = "Resources/lending_data.csv"
len_df = pd.read_csv(path)
len_df.tail()

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

In [8]:
#To get statictical summary of the data set
len_df.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

# Prediction:
I do believe the Logistic Regression will perform better than the Random Forests Classifier. 

In [9]:
# Splitting the data into X_train, X_test, y_train, y_test
X = len_df.drop("loan_status", axis=1 )
y = len_df["loan_status"]

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)

In [10]:
# To Scale the data provided
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
scaler = StandardScaler().fit(X_test)
X_test_scaled = scaler.transform(X_test)

In [11]:
# Logistic Regression model print the model score
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression().fit(X_train_scaled, y_train)
print(f"Training Score: {classifier.score(X_train_scaled, y_train)}")
print(f"Test Score: {classifier.score(X_test_scaled, y_test)}")

Training Score: 0.9941188609162196
Test Score: 0.9941704498555509


In [12]:
#Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state = 42, n_estimators = 500)
clf.fit(X_train_scaled, y_train)
print(f"Training Score: {clf.score(X_train_scaled, y_train)}")
print(f"Test Score: {clf.score(X_test_scaled, y_test)}")

Training Score: 0.9971970009629936
Test Score: 0.9929323153115972


In [13]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state = 42, n_estimators = 500)
clf.fit(X_train_scaled, y_train)
print(f"Training Score: {clf.score(X_train_scaled, y_train)}")
print(f"Test Score: {clf.score(X_test_scaled, y_test)}")

Training Score: 0.9971970009629936
Test Score: 0.9929323153115972


# Review/Conclusion

Based on the above, the random forest testing score 0.99417 is higher than the logistic Regression testing score of 0.9929, which is not same with the earlier prediction. But both model are performing well.